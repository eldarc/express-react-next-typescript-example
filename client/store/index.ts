import { useStaticRendering } from 'mobx-react'
import AuthStore from './modules/AuthStore'
import UserStore from './modules/UserStore'
import OverviewStore from './modules/OverviewStore'
import Cookies from 'js-cookie'

// For SSR static rendering is required.
const isServer = (typeof window === 'undefined')
useStaticRendering(isServer)

// Main root store.
class Store {
  // @ts-ignore (AuthStore is definitely set in the `else` branch in the constructor)
  public auth: AuthStore
  public user: UserStore
  public overview: OverviewStore

  constructor(checkAuth: boolean) {
    this.overview = new OverviewStore()
    this.user = new UserStore()

    // Check if the user is already logged in, and create the store according to that status.
    const token = Cookies.get('token')
    if (checkAuth && token) {
      this.auth = new AuthStore(token)

      // Fetch user data.
      this.user.fetchUserData().then(() => {
        this.overview.ignoreUser(this.user.userData._id)
      })
    } else {
      this.auth = new AuthStore()
    }
  }
}

// Export Store and store initialization function.
export { Store }
let store: Store
export function initializeStore() {
  // Always make a new store if server, otherwise state is shared between requests
  if (isServer) {
    // return new Store(initialData)
    return new Store(false)
  }

  if (!store) {
    store = new Store(true)
  }

  return store
}
