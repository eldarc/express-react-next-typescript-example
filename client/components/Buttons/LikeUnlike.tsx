import * as React from 'react'
import { observer, inject } from 'mobx-react'

// ~~~ Layout ~~~
import { Button, Spinner } from 'reactstrap'
import { Store } from '@/store'

// ~~~ Interfaces ~~~
interface IProps {
  store: Store
  userId: string
  mode: string
}

interface IState {
  processing: boolean
}

// ~~~ Component ~~~
@inject('store')
@observer
class LikeUnlike extends React.Component<IProps, IState> {
  constructor(props: IProps) {
    super(props)

    this.state = {
      processing: false
    }
  }

  public likeOrUnlike = () => {
    this.props.mode === 'like' ? this.likeUser() : this.unlikeUser()
  }

  public likeUser = async () => {
    try {
      await this.setState({
        processing: true
      })

      await this.props.store.overview.doLikeUser(this.props.userId)
    } catch (e) {
      console.log(e)
    } finally {
      this.setState({
        processing: false
      })
    }
  }
  public unlikeUser = async () => {
    try {
      await this.setState({
        processing: true
      })

      await this.props.store.overview.doUnlikeUser(this.props.userId)
    } catch (e) {
      console.log(e)
    } finally {
      this.setState({
        processing: false
      })
    }
  }

  public render() {
    return (
      <Button className="like-unlike-btn" onClick={this.likeOrUnlike} color={this.props.mode === 'like' ? 'success' : 'danger'} size="sm">
        {
          this.state.processing ? <Spinner size="sm" color="light" />
          : (this.props.mode === 'like' ? 'Like' : 'Unlike')
        }
      </Button>
    )
  }
}

export default LikeUnlike
