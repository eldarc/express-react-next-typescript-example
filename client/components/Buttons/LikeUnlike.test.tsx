/* eslint-env jest */
import React from 'react'
import renderer from 'react-test-renderer'
import { Store } from '@/store'

import LikeUnlike from './LikeUnlike'

describe('LikeUnlike', () => {
  it('renders correctly like mode', () => {
    const tree = renderer
        .create(<LikeUnlike mode="like" userId="1" store={new Store(false)}/>)
        .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders correctly unlike mode', () => {
    const tree = renderer
        .create(<LikeUnlike mode="unlike" userId="1" store={new Store(false)}/>)
        .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
