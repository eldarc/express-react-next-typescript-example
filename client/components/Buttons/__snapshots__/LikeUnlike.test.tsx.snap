// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LikeUnlike renders correctly like mode 1`] = `
<button
  aria-label={null}
  className="like-unlike-btn btn btn-success btn-sm"
  onClick={[Function]}
  type="button"
>
  Like
</button>
`;

exports[`LikeUnlike renders correctly unlike mode 1`] = `
<button
  aria-label={null}
  className="like-unlike-btn btn btn-danger btn-sm"
  onClick={[Function]}
  type="button"
>
  Unlike
</button>
`;
